general/networking/nginx.md

```diff
@@ -10,65 +10,87 @@ title: Nginx Reverse Proxy
     # return 301 https://$host$request_uri;
-}
+#}
 
-# Uncomment this section after you have acquired a SSL Certificate
 #server {
-#    listen 443 ssl http2;
-#    server_name DOMAIN_NAME;
-#    ssl_certificate /etc/letsencrypt/live/DOMAIN_NAME/fullchain.pem;
-#    ssl_certificate_key /etc/letsencrypt/live/DOMAIN_NAME/privkey.pem;
-#    include /etc/letsencrypt/options-ssl-nginx.conf;
-#    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;
-#    add_header Strict-Transport-Security "max-age=31536000" always;
-#    ssl_trusted_certificate /etc/letsencrypt/live/DOMAIN_NAME/chain.pem;
-#    ssl_stapling on;
-#    ssl_stapling_verify on;
-#
-#    # Security / XSS Mitigation Headers
-#    add_header X-Frame-Options "SAMEORIGIN";
-#    add_header X-XSS-Protection "1; mode=block";
-#    add_header X-Content-Type-Options "nosniff";
-#
-#    # Content Security Policy
-#    # See: https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP
-#    # Enforces https content and restricts JS/CSS to origin
-#    # External Javascript (such as cast_sender.js for Chromecast) must be whitelisted.
-#    add_header Content-Security-Policy "default-src https: data: blob:; style-src 'self' 'unsafe-inline'; script-src 'self' 'unsafe-inline' https://www.gstatic.com/cv/js/sender/v1/cast_sender.js; worker-src 'self' blob:; connect-src 'self'; object-src 'none'; frame-ancestors 'self'";
-#
-#    location / {
-#        # Proxy main Jellyfin traffic
-#        proxy_pass http://SERVER_IP_ADDRESS:8096/;
-#        proxy_set_header Host $host;
-#        proxy_set_header X-Real-IP $remote_addr;
-#        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
-#        proxy_set_header X-Forwarded-Proto $scheme;
-#        proxy_set_header X-Forwarded-Protocol $scheme;
-#        proxy_set_header X-Forwarded-Host $http_host;
-#
-#        # Disable buffering when the nginx proxy gets very resource heavy upon streaming
-#        proxy_buffering off;
-#    }
-#    location /socket {
-#        # Proxy Jellyfin Websockets traffic
-#        proxy_pass http://SERVER_IP_ADDRESS:8096/socket;
-#        proxy_http_version 1.1;
-#        proxy_set_header Upgrade $http_upgrade;
-#        proxy_set_header Connection "upgrade";
-#        proxy_set_header Host $host;
-#        proxy_set_header X-Real-IP $remote_addr;
-#        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
-#        proxy_set_header X-Forwarded-Proto $scheme;
-#        proxy_set_header X-Forwarded-Protocol $scheme;
-#        proxy_set_header X-Forwarded-Host $http_host;
-#    }
-#}
+    # listen 443 ssl http2;
+    # listen [::]:443 ssl http2;
+    server_name DOMAIN_NAME;
+
+    # use a variable to store the upstream proxy
+    # in this example we are using a hostname which is resolved via DNS
+    # (if you aren't using DNS remove the resolver line and change the variable to point to an IP address e.g `set $jellyfin 127.0.0.1`)
+    set $jellyfin jellyfin;
+    resolver 127.0.0.1 valid=30;
```

Comment:
I'll be honest, I've never actually seen the benefit of doing this over just `proxy_pass http://$ip_address` or `proxy_pass http://$hostname` unless you're using nginx as a load balancer.  Just seems like more lines of code to accomplish the same thing.  I'm not opposed, I just want to understand why this is preferred.

Reply:
This is just my personal preference but I prefer using variables like `set $jellyfin IP_ADDRESS_OR_HOSTNAME;` because it makes it easier to update in the future (you don't have to change the IP address or hostname in multiple places just at the top in one place). It also means this config can mostly be used as is just by updating the variable to point to the right IP or hostname and removing the resolver line if that's not necessary (it's necessary in my case because I'm using an lxc container without a static IP so everything is dynamic).

Reply:
You should do this with an [upstream](http://nginx.org/en/docs/http/ngx_http_upstream_module.html#upstream):

```
upstream jellyfin {
	server jellyfin:8096;
}
server {
	...
	location / {
		proxy_pass http://jellfyin;
	}
}
```

Using variables is [discouraged](http://nginx.org/en/docs/faq/variables_in_config.html):

> Using variables to store static strings is also a bad idea.